### DISCUSSION TITLE
Gracefully handling database connection errors

### DISCUSSION THREAD

**Discussion by fiznool** [timgit/pg-boss#245] [Q&A]
We are running pg-boss (very successfully) on Heroku, using Heroku Postgres. Every now and again, Heroku performs scheduled maintenance on our database, which takes the DB offline for a short while.

We've set up our server to use an existing knex adapter, meaning that we have implemented `executeSql` to run the sql through `knex.raw`. We've also set up a global error handler to catch any emitted errors on the pgBoss instance, currently this logs out the error and then re-throws it.

Here's how it currently looks:

``` js
import knex from './knex.provider';

const boss = new PgBoss({
  // Funnel the SQL through knex, so that the global connection pool parameters are respected.
  // https://github.com/timgit/pg-boss/blob/master/docs/configuration.md#database-options
  db: {
    executeSql(sql, params = []) {
      // pg-boss uses `$1, $2, $n` notation to denote parameters, whereas knex uses `:name` notation.
      // We need to convert the SQL string to knex notation.
      // Let's do this with a `:p` prefix. So, `$1` becomes `:p1`, `$2` becomes `:p2`, etc.
      const knexSql = sql.replace(
        /\$(\d+)/g,
        (_, digit) => `:p${digit}`
      );

      // pg-boss passes an array of parameters, which are used to replace the associated `$n` definition in the SQL.
      // knex needs an object of key-value pairs to achieve the same result, where the key corresponds to the `:name` in the SQL string.
      // As per the above, we change `$1` into `:p1`, so we must convert the parameters into a map,
      // where each key takes the form `:p{x}` where `{x}` is the same number as used in the original `${x}` notation.
      const knexParams = params.reduce(
        (acc, val, idx) => ({
          ...acc,
          // The parameter indexes start at 1 instead of 0.
          // In addition, any undefined values must be explicitly set to `null`,
          // as knex does not recognise `undefined` as a valid parameter value.
          [`p${idx + 1}`]: typeof val !== 'undefined' ? val : null,
        }),
        {}
      );

      // Finally, run the query through knex.
      return knex.raw(knexSql, knexParams);
    },
  },
});

boss.on('error', (error) => {
  console.log(error);
  throw error;
});
```

The issue we're having is that when the database is unavailable, the `on('error'` listener is catching an error when a worker tries to pull down a job from the db. Since we have code to re-throw the error, this bubbles up and crashes the server. We'd like to gracefully handle this error, so that the server does not need to be restarted in this case.

Here's an example of the error we get:

``` js
{
  "type": "Error",
  "message": "connect ECONNREFUSED 172.31.0.2:5432",
  "stack":
      Error: connect ECONNREFUSED 172.31.0.2:5432
          at Object.raw (/app/node_modules/knex/lib/util/make-knex.js:114:30)
          at Function.raw (/app/node_modules/knex/lib/util/make-knex.js:90:29)
          at Object.executeSql (/app/server/providers/pg-boss.provider.ts:68:19)
          at Manager.fetch (/app/node_modules/pg-boss/src/manager.js:244:34)
          at Object.fetch (/app/node_modules/pg-boss/src/manager.js:101:25)
          at Worker.start (/app/node_modules/pg-boss/src/worker.js:12:25)
  "errno": -111,
  "code": "ECONNREFUSED",
  "syscall": "connect",
  "address": "172.31.0.2",
  "port": 5432,
  "originalStack": "Error: connect ECONNREFUSED 172.31.0.2:5432\n    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1146:16)"
}
```

My first thought was to add some additional logic in the `on('error'` listener, to check if the error has the code of `ECONNREFUSED` - if this is the case, we would log out the error, but not re-throw it, meaning that the server would not crash.

My question - is this the recommended way to handle this sort of thing? My other thought was to add a `catch` to the `knex.raw` call inside `executeSql`, and return an artificially generated knex 'empty response' so that pg-boss thinks that no rows were returned. But I figured this is more error prone, and could have a knock-on effect elsewhere?

**Comment by timgit** [comment]
Adding this logic inside your custom db adapter is probably the best way to handle this specific case in the worker, but other db use cases may require different "empty responses" and still fail and throw. 

Does Heroku provide a stop signal ahead of scheduled maintenance? If so, you could use that to gracefully stop pg-boss during the maintenance. You will need to use pg-boss v6 or higher for that. This is currently only published in beta, so let me know if that prevents you from using it. I am planning on merging this early next week. 

Hope this helps!

**Reply by fiznool** [reply]
Thank you for your prompt reply. This is very helpful indeed.

Your comment has prompted us to rethink the approach to maintenance and I think we will carry out the maintenance manually, so we can shutdown the queues manually before the database goes offline.

It has however brought up a larger point around error handling. We've identified two different strategies for handling errors, depending on the type of error that has occurred:

- Fatal: these are if (e.g.) the database can't be contacted when pg-boss attempts to migrate its schema.
- Recoverable: these are if (e.g.) the database can't be contacted when a worker polls for a job

It would be nice if we could distinguish between these types of errors, so the error handling logic can know to either a) crash and restart the server, or b) log a warning but continue.

It would be great to hear your thoughts on whether this is a good idea or not?

**Reply by timgit** [reply]
Migration errors are promise rejections on start(), and all other errors emit `error` and will keep going.  Are you asking about a different case I'm missing?
